### PR TITLE
admin: views for template definitions and counters

### PR DESCRIPTION
**Pull request by omelkonian**
- Adds admin views for management of TemplateDefinition and Counter.
- Adds a Jinja template for the admin interface, which provides a reset
  row action.

Signed-off-by: Orestis Melkonian melkon.or@gmail.com

### CODE REVIEW

File: invenio_sequencegenerator/admin.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        'template_instance',
+        'counter',
+        'definition_name',
+    )
+
+    @action('reset', 'Reset',
+            'Are you sure you want to reset selected counter?')
+    def reset(self, ids):
+        """Reset selected counter."""
+        for id in ids:
```
Comment (PXke) [inline invenio_sequencegenerator/admin.py]: id is a function of python you should try to avoid to use variable with the same name, even if it doesn't break anything here.

File: invenio_sequencegenerator/models.py [hunk 2]
```diff
@@ -27,15 +27,15 @@
 from sqlalchemy.orm.exc import MultipleResultsFound
 
 from .errors import InvalidResetCall, InvalidTemplate
-from .utils import double_counter, extract_placeholders
+from .utils import double_counter, extract_placeholders, undouble_counter
 
 
 class TemplateDefinition(db.Model, object):
     """Representation of a template definition."""
```
Comment (PXke) [inline invenio_sequencegenerator/models.py:34]: Could you please add more details here ? I didn't find anything in the documentation about it.

Comment (jirikuncar) [inline invenio_sequencegenerator/models.py:34]: @PXke you are not suppose to use this one. Check the `api.py`.

File: invenio_sequencegenerator/admin.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        flash('{0} successfully reset.'.format(
+            'Counters were' if len(ids) > 1 else 'Counter was'
+        ))
+        return redirect(url_for('counter.index_view'))
+
+
+templatedefinition_adminview = {
+    'model': TemplateDefinition,
+    'modelview': TemplateDefinitionView,
+    'category': _('Template Definition Management')
```
Comment (jirikuncar) [inline invenio_sequencegenerator/admin.py]: both views should be under the same category e.g. Sequences

File: invenio_sequencegenerator/models.py [hunk 4]
```diff
@@ -102,6 +102,16 @@ def counter(self, **kwargs):
 
         return counter
 
+    def __repr__(self):
+        """Canonical representation of ``TemplateDefinition``."""
+        new_template = undouble_counter(self.meta_template, self.COUNTER_REGEX)
+        return ('TemplateDefinition('
+                'name="{0.name}", '
```
Comment (jirikuncar) [inline invenio_sequencegenerator/models.py]: use `!repr` instead `"..."`

File: invenio_sequencegenerator/models.py (hunk 2 above)
Comment (PXke) [inline invenio_sequencegenerator/models.py:34]: @jirikuncar  Yes, I understand but if I understood correctly  from : 

> +.. automodule:: invenio_sequencegenerator.models
> -   :members:

The documentation will generate itself from the comments, I really find this docstring not helpful to understand the code :/

File: invenio_sequencegenerator/models.py (hunk 4 above)
Comment (PXke) [inline invenio_sequencegenerator/models.py]: Any link for that I don't understand it ?

Comment (jirikuncar) [inline invenio_sequencegenerator/models.py]: https://pyformat.info/ - Value conversion

**Comment by omelkonian** [comment]
@jirikuncar What's wrong with coveralls?? No report yet.

**Comment by jirikuncar** [comment]
@omelkonian I don't know :( Anyway, can you rebase your branch?

File: examples/app.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+# Create admin
+admin = admin.Admin(app, name='Example: Sequence Generator',
+                    template_mode='bootstrap3')
+
+# Add views
+admin.add_view(ta['modelview'](ta['model'], db.session))
+admin.add_view(ca['modelview'](ca['model'], db.session))
+
+
+if __name__ == '__main__':
```
Comment (jirikuncar) [inline examples/app.py]: please remove this part as it's not needed since we have `flask run` command.

File: invenio_sequencegenerator/admin.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    column_list = (
+        'template_instance',
+        'counter',
+        'definition_name',
+    )
+
+    @expose('/reset', methods=('POST',))
+    def reset_view(self):
+        """Reset selected counter."""
+        start = request.form['start']
```
Comment (jirikuncar) [inline invenio_sequencegenerator/admin.py]: `start = request.form.get('start', type=int, 0)`

File: invenio_sequencegenerator/templates/invenio_sequencegenerator/custom_list.html [hunk 7]
```diff
@@ -0,0 +1,13 @@
+{% extends 'admin/model/list.html' %}
```
Comment (jirikuncar) [inline invenio_sequencegenerator/templates/invenio_sequencegenerator/custom_list.html]: missing license header

File: tests/test_admin.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2015, 2016 CERN.
```
Comment (jirikuncar) [inline tests/test_admin.py]: only 2016

File: tests/test_models.py [hunk 9]
```diff
@@ -238,7 +238,28 @@ def test_regex():
...
 def test_py3_next(db):
     """Test PY3 __next__."""
     assert Sequence(Template.create('PY3', '{counter}')).__next__() == '0'
+
+
+def test_repr(db):
+    """Test __repr__."""
+    template = Template.create('A', '{counter}')
+    assert repr(template.model) == (
+        'TemplateDefinition(name=\'A\', '
```
Comment (jirikuncar) [inline tests/test_models.py]: in this case you can use double quotes: `"TemplateDefinition(name='A', "`

**Comment by omelkonian** [comment]
ping @jirikuncar

File: examples/app.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    db.session.commit()
+
+
+@app.route('/')
+def index():
+    return '<a href="/admin/">Click me to get to Admin!</a>'
+
+
+# Create admin
+admin = admin.Admin(app, name='Example: Sequence Generator',
```
Comment (jirikuncar) [inline examples/app.py]: Try to avoid overriding global imports. What about following pattern:

``` python
from flask_admin import Admin

admin = Admin(...)
```

or

``` python
import flask_admin

admin = flask_admin.Admin(...)
```

File: invenio_sequencegenerator/admin.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        template_instance = request.form['rowid']
+        Counter.query.get(template_instance).reset(start=start)
+        db.session.commit()
+        return redirect(url_for('{0}.index_view'.format(self.endpoint)))
+
+
+templatedefinition_adminview = {
+    'model': TemplateDefinition,
+    'modelview': TemplateDefinitionView,
+    'category': _('Sequences')
```
Comment (jirikuncar) [inline invenio_sequencegenerator/admin.py]: please add comma at the end of line

File: invenio_sequencegenerator/admin.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    )
+
+    @expose('/reset', methods=('POST',))
+    def reset_view(self):
+        """Reset selected counter."""
+        start = request.form.get('start', default=0, type=int)
+        template_instance = request.form['rowid']
+        Counter.query.get(template_instance).reset(start=start)
+        db.session.commit()
+        return redirect(url_for('{0}.index_view'.format(self.endpoint)))
```
Comment (jirikuncar) [inline invenio_sequencegenerator/admin.py]: or you can just use `url_for('.index_view')`

File: invenio_sequencegenerator/models.py [hunk 13]
```diff
@@ -63,13 +66,12 @@ def validate_meta_template(self, key, value):
         """Validate template string of template definition."""
         if not self.COUNTER_REGEX.search(value):
             raise InvalidTemplate('No counter placeholder')
-        value = double_counter(value, self.COUNTER_REGEX)
         return value
 
-    @hybrid_property
+    @property
```
Comment (jirikuncar) [inline invenio_sequencegenerator/models.py]: why did you change the hybrid property to simple property?

File: invenio_sequencegenerator/templates/invenio_sequencegenerator/custom_list.html [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+  {{ super() }}
+  <form class="icon" method="POST" action="{{ url_for('.reset_view') }}">
+    <input name="start" value="0" type="number" size="1">
+    <input name="rowid" value="{{ get_pk_value(row) }}" type="hidden">
+    <button onclick="return confirm('Are you sure you want to reset selected counter?');" title="Reset">
+      <span class="fa fa-step-backward glyphicon glyphicon-step-backward">
+      </span>
+    </button>
+  </form>
+{% endblock %}
```
Comment (jirikuncar) [inline invenio_sequencegenerator/templates/invenio_sequencegenerator/custom_list.html]: missing new line

File: invenio_sequencegenerator/models.py (hunk 13 above)
Comment (omelkonian) [inline invenio_sequencegenerator/models.py]: After adding `undouble_bracket(....)` it did not work with `@hybrid_property`

Comment (omelkonian) [inline invenio_sequencegenerator/models.py]: From what I understood, if it was a `@hybrid_property`, it would require translation to an SQL expression when used in a query or sth like that. but it uses the `undouble_counter` method that uses regex and this cannot be translated (at least automatically) to SQL.

Comment (jirikuncar) [inline invenio_sequencegenerator/models.py]: Do you have the exception?

Comment (omelkonian) [inline invenio_sequencegenerator/models.py]: Let me recreate it.

Comment (omelkonian) [inline invenio_sequencegenerator/models.py]:
``` python
    def test_iter(db):
        """Test simple files use-case."""
>       fl = Sequence(Template.create('FL', 'File {counter}'))

tests/test_models.py:176: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
invenio_sequencegenerator/api.py:56: in create
    name=name, meta_template=meta_template, start=start, step=step
<string>:4: in __init__
    ???
../local/lib/python2.7/site-packages/sqlalchemy/orm/state.py:306: in _initialize_instance
    manager.dispatch.init_failure(self, args, kwargs)
../local/lib/python2.7/site-packages/sqlalchemy/util/langhelpers.py:60: in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
../local/lib/python2.7/site-packages/sqlalchemy/orm/state.py:303: in _initialize_instance
    return manager.original_init(*mixed[1:], **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <[TypeError("expected string or buffer") raised in repr()] SafeRepr object at 0x7ff8873487a0>
kwargs = {'meta_template': 'File {counter}', 'name': 'FL', 'start': 0, 'step': 1}
cls_ = <class 'invenio_sequencegenerator.models.TemplateDefinition'>, k = 'meta_template'

    def _declarative_constructor(self, **kwargs):
        """A simple constructor that allows initialization from kwargs.

        Sets attributes on the constructed instance using the names and
        values in ``kwargs``.

        Only keys that are present as
        attributes of the instance's class are allowed. These could be,
        for example, any mapped columns or relationships.
        """
        cls_ = type(self)
        for k in kwargs:
            if not hasattr(cls_, k):
                raise TypeError(
                    "%r is an invalid keyword argument for %s" %
>                   (k, cls_.__name__))
E               TypeError: 'meta_template' is an invalid keyword argument for TemplateDefinition

../local/lib/python2.7/site-packages/sqlalchemy/ext/declarative/base.py:649: TypeError

```

Comment (jirikuncar) [inline invenio_sequencegenerator/models.py]: What about moving counter escaping to the `Counter` class where it's needed?

File: setup.py [hunk 15]
```diff
@@ -107,6 +108,12 @@
             'invenio_sequencegenerator = '
             'invenio_sequencegenerator.models',
         ],
+        'invenio_admin.views': [
+            'invenio_sequencegenerator_user = '
```
Comment (jirikuncar) [inline setup.py]: `invenio_sequencegenerator_templatedefinition` ?

## 

copy&paste problem?

File: setup.py [hunk 16]
```diff
@@ -107,6 +108,12 @@
...
+            'invenio_sequencegenerator.admin:templatedefinition_adminview',
+            'invenio_sequencegenerator_role = '
```
Comment (jirikuncar) [inline setup.py]: `_counter`

File: invenio_sequencegenerator/admin.py [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+
+    list_template = 'invenio_sequencegenerator/custom_list.html'
+
+    column_list = (
+        'template_instance',
+        'counter',
+        'definition_name',
+    )
+
+    @expose('/reset', methods=('POST',))
```
Comment (omelkonian) [inline invenio_sequencegenerator/admin.py:60]: Should I make it `@staticmethod` as the QuantifiedCode check suggests?

Comment (jirikuncar) [inline invenio_sequencegenerator/admin.py:60]: Not for now.

File: invenio_sequencegenerator/templates/invenio_sequencegenerator/custom_list.html [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.}
+#}
+{% extends 'admin/model/list.html' %}
```
Comment (jirikuncar) [inline invenio_sequencegenerator/templates/invenio_sequencegenerator/custom_list.html]: please strip whitespaces with `{%-`

File: setup.py [hunk 19]
```diff
@@ -40,6 +40,7 @@
     'pytest-cov>=1.8.0',
     'pytest-pep8>=1.0.6',
     'pytest>=2.8.0',
+    'Flask-Admin>=1.4.2',
```
Comment (jirikuncar) [inline setup.py]: please sort the imports

File: examples/app.py [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+from flask_babelex import Babel
+from flask_cli import FlaskCLI
+from invenio_db import InvenioDB, db
+from invenio_sequencegenerator.admin import counter_adminview as ca
+from invenio_sequencegenerator.admin import templatedefinition_adminview as ta
+from invenio_sequencegenerator.api import Sequence, Template
+
+# Create Flask application
+app = Flask(__name__,
+            template_folder='../invenio_sequencegenerator/templates/')
```
Comment (jirikuncar) [inline examples/app.py]: we might need to create a dummy blueprint which loads the templates

File: MANIFEST.in [hunk 21]
```diff
@@ -46,3 +46,5 @@ recursive-include invenio_sequencegenerator *.po
 recursive-include invenio_sequencegenerator *.pot
 recursive-include invenio_sequencegenerator *.py
 recursive-include tests *.py
+recursive-include examples *.py
+recursive-include invenio_sequencegenerator *.html
```
Comment (jirikuncar) [inline MANIFEST.in]: resort

**Comment by omelkonian** [comment]
ping @jirikuncar

File: invenio_sequencegenerator/ext.py [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+        if app:
+            self.init_app(app)
+
+    def init_app(self, app):
+        """Flask application initialization."""
+        self.init_config(app)
+        app.extensions['invenio-sequencegenerator'] = self
+
+        # Register dummy blueprint just for loading templates
+        template_loader = Blueprint('template_loader', __name__,
```
Comment (jirikuncar) [inline invenio_sequencegenerator/ext.py]: - `invenio_sequencegenerator` instead of `template_loader`
- you can do it in one go:

``` python
app.register_blueprint(Blueprint(
    'invenio_sequencegenerator', ...
)))
```

File: invenio_sequencegenerator/ext.py [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+                                    template_folder='templates')
+        app.register_blueprint(template_loader)
+
+    def init_config(self, app):
+        """Initialize configuration."""
+        app.config.setdefault(
+            'SEQUENCEGENERATOR_ADMIN_TEMPLATE',
```
Comment (jirikuncar) [inline invenio_sequencegenerator/ext.py]: do you need it?

Comment (omelkonian) [inline invenio_sequencegenerator/ext.py]: not really :)

File: invenio_sequencegenerator/admin.py (hunk 11 above)
Comment (jirikuncar) [inline invenio_sequencegenerator/admin.py]: ping @omelkonian

**Comment by omelkonian** [comment]
@jirikuncar ping

**Comment by jirikuncar** [comment]
@omelkonian please rebase the branch. Thanks

File: invenio_sequencegenerator/admin.py (hunk 11 above)
Comment (jirikuncar) [reply invenio_sequencegenerator/admin.py]: @omelkonian have you seen this comment?

File: examples/app.py [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+from invenio_sequencegenerator.ext import InvenioSequenceGenerator
+
+app = Flask(__name__)
+app.config.update(
+    SECRET_KEY="CHANGE_ME",
+    SECURITY_PASSWORD_SALT="CHANGE_ME_ALSO",
+    SQLALCHEMY_DATABASE_URI=os.environ.get(
+        'SQLALCHEMY_DATABASE_URI', 'sqlite:///test.db'),
+    SQLALCHEMY_TRACK_MODIFICATIONS=True,
+    TESTING=True,
```
Comment (jirikuncar) [inline examples/app.py]: do you need `TESTING=True` for an example app?

File: invenio_sequencegenerator/templates/invenio_sequencegenerator/custom_list.html [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.}
+#}
+{%- extends 'admin/model/list.html' %}
+{%- block list_row_actions %}
+  {{ super() }}
+  <form class="icon" method="POST" action="{{ url_for('.reset_view') }}">
+    <input name="start" value="0" type="number" size="1">
+    <input name="rowid" value="{{ get_pk_value(row) }}" type="hidden">
+    <button onclick="return confirm('Are you sure you want to reset selected counter?');" title="Reset">
```
Comment (jirikuncar) [inline invenio_sequencegenerator/templates/invenio_sequencegenerator/custom_list.html]: please make sure that all strings are marked for translations with `{{ _('...') }}`

File: invenio_sequencegenerator/templates/invenio_sequencegenerator/custom_list.html [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+      <span class="fa fa-step-backward glyphicon glyphicon-step-backward">
```
Comment (jirikuncar) [inline invenio_sequencegenerator/templates/invenio_sequencegenerator/custom_list.html]: can we use `<i class=""></i>` instead of `<span` for icons?

File: invenio_sequencegenerator/templates/invenio_sequencegenerator/custom_list.html [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+      </span>
```
Comment (jirikuncar) [inline invenio_sequencegenerator/templates/invenio_sequencegenerator/custom_list.html]: there should not be any whitespace inside icon `<span ...></span>`

File: invenio_sequencegenerator/admin.py (hunk 11 above)
Comment (jirikuncar) [reply invenio_sequencegenerator/admin.py]: @omelkonian ping

File: invenio_sequencegenerator/ext.py [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+        if app:
+            self.init_app(app)
+
+    def init_app(self, app):
+        """Flask application initialization."""
+        app.extensions['invenio-sequencegenerator'] = self
+
+        # Register dummy blueprint just for loading templates
+        app.register_blueprint(Blueprint(
+            'template_loader', __name__, template_folder='templates')
```
Comment (jirikuncar) [inline invenio_sequencegenerator/ext.py]: please use `invenio_sequencegenerator` instead of `template_loader`

File: invenio_sequencegenerator/models.py [hunk 29]
```diff
@@ -102,6 +81,33 @@ def counter(self, **kwargs):
...
+    """Automatically derive parent from template string."""
+    placeholders = extract_placeholders(value)
+
+    # Check if parent sequence exists
+    try:
+        parent = TemplateDefinition.query.filter(
+            TemplateDefinition.name.in_(placeholders)
+        ).one_or_none() if placeholders else None
+    except MultipleResultsFound:
+        raise InvalidTemplate('More than 1 parents in template')
```
Comment (jirikuncar) [inline invenio_sequencegenerator/models.py]:
``` python
raise InvalidTemplate(
    'More than 1 parents in template "{0}".'.format(value)
)
```

**Review by jirikuncar — CHANGES_REQUESTED** [review]

**Comment by omelkonian** [comment]
@jirikuncar ping

**Review by jirikuncar — APPROVED** [review]

File: invenio_sequencegenerator/admin.py [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Admin views for invenio-accounts."""
+
+from flask import flash, redirect, request, url_for
+from flask_admin import expose
+from flask_admin.actions import action
```
Comment (jirikuncar) [inline invenio_sequencegenerator/admin.py:29]: unused

File: invenio_sequencegenerator/admin.py [hunk 31]
```diff
@@ -0,0 +1,82 @@
...
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Admin views for invenio-accounts."""
+
+from flask import flash, redirect, request, url_for
```
Comment (jirikuncar) [inline invenio_sequencegenerator/admin.py:27]: `flash` is not used

File: examples/app.py [hunk 32]
```diff
@@ -0,0 +1,104 @@
...
+    fl15.next()
+    fl15.next()
+    fl16 = Sequence(fl, PL='2016: Playlist 1')
+    fl16.next()
+    db.session.commit()
+
+
+@app.route('/')
+def index():
+    return '<a href="/admin/">Click me to get to Admin!</a>'
```
Comment (jirikuncar) [inline examples/app.py:95]: missing docstring

**Review by jirikuncar — CHANGES_REQUESTED** [review]
Please fix also copyright `2016 C` to `2016 CERN.` in `tests/test_models.py`. Thanks
